azure-webapp-maven-plugin/src/main/java/com/microsoft/azure/maven/webapp/WebAppUtils.java

```diff
@@ -42,14 +55,88 @@ public static void assureWindowsWebApp(final WebApp app) throws MojoExecutionExc
         }
     }
 
-    public static WithNewAppServicePlan defineApp(final AbstractWebAppMojo mojo) throws Exception {
-        final WithGroup<WithNewAppServicePlan> withGroup = mojo.getAzureClient().webApps()
+    public static WithDockerContainerImage defineLinuxApp(final AbstractWebAppMojo mojo, final AppServicePlan plan)
+            throws Exception {
+        final String resourceGroup = mojo.getResourceGroup();
+
+        assureLinuxPlan(plan);
+
+        final ExistingLinuxPlanWithGroup existingLinuxPlanWithGroup = mojo.getAzureClient().webApps()
                 .define(mojo.getAppName())
-                .withRegion(mojo.getRegion());
+                .withExistingLinuxPlan(plan);
+        return mojo.getAzureClient().resourceGroups().contain(resourceGroup) ?
+                existingLinuxPlanWithGroup.withExistingResourceGroup(resourceGroup) :
+                existingLinuxPlanWithGroup.withNewResourceGroup(resourceGroup);
+    }
+
+    private static void assureLinuxPlan(final AppServicePlan plan) throws MojoExecutionException {
+        if (!plan.operatingSystem().equals(OperatingSystem.LINUX)) {
+            throw new MojoExecutionException(String.format(SERVICE_PLAN_NOT_APPLICABLE,
+                    plan.name(), OperatingSystem.LINUX.name()));
+        }
+    }
+
+    public static WithCreate defineWindowsApp(final AbstractWebAppMojo mojo, final AppServicePlan plan)
```

Comment:
seems the logic is same except for `assurePlan`? is it possible to extract the same logic to one method?

Reply:
Emm, besides the `assurePlan()`, their return types are also different. Any suggestions?

Reply:
could the return type be a base type?

Reply:
Seems could not.

- WithDockerContainerImage: https://github.com/Azure/azure-libraries-for-java/blob/master/azure-mgmt-appservice/src/main/java/com/microsoft/azure/management/appservice/WebApp.java#L240
- WithCreate: https://github.com/Azure/azure-libraries-for-java/blob/master/azure-mgmt-appservice/src/main/java/com/microsoft/azure/management/appservice/WebApp.java#L302

Reply:
i see...